api-module-library/frontify/api.js

```diff
@@ -333,6 +334,64 @@ class Api extends OAuth2Requester {
             assets: response.data.brand.search.edges,
         };
     }
+
+    async createAsset(asset) {
+        const ql = `mutation CreateAsset {
+                      createAsset(input: {
+                        fileId: "${asset.id}",
+                        title: "${asset.title}",
+                        projectId: "${asset.projectId}"
+                      }) {
+                        job {
+                          assetId
+                        }
+                      }
+                    }`;
+
+        const response = await this._post(this.buildRequestOptions(ql));
+        this.assertResponse(response);
+        return {
+            id: response.data.createAsset.job.assetId
+        };
+    }
+
+    async createFileId(input) {
+        const ql = `mutation UploadFile {
+                      uploadFile(input: {
+                        filename: "${input.filename}",
+                        size: ${input.size},
+                        chunkSize: ${input.chunkSize}
+                      }) {
+                        id
+                        urls
+                      }
+                    }`;
+
+        const response = await this._post(this.buildRequestOptions(ql));
+        this.assertResponse(response);
+        return response.data.uploadFile;
+    }
+
+    // Total of addresses in urls should match the number of chunks in
+    // stream. The code invoking this method should take care of this
+    // using a correct "highWaterMark".
+    async uploadFile(stream, urls) {
+        for await (const chunk of stream) {
+            // AWS url
+            const url = urls.shift();
+
+            // Using fetch to avoid sending Frontify auth headers to AWS
+            await fetch(url, {
+                method: 'PUT',
+                headers: {
+                    'content-type': 'binary'
+                },
+                body: chunk
+            });
+        }
+
+        return Promise.resolve();
```

Comment:
@roboli Instead of just resolving the promise, it would be better to return the results from Frontify's call, or something along those lines.
What is the API return?

Reply:
Actually, is to AWS to which the file is uploaded to, not Frontify. I followed their example which ended returning a resolved promise. But I could change returning last chunk response?

Reply:
https://developer.frontify.com/d/XFPCrGNrXQQM/graphql-api#/deep-dive/upload-file-create-asset

Reply:
This is what AWS is returning after uploading a file chunk @leofmds:

```
[
  Response {
    size: 0,
    timeout: 0,
    [Symbol(Body internals)]: { body: [PassThrough], disturbed: false, error: null },
    [Symbol(Response internals)]: {
      url: 'https://s3.amazonaws.com/frontify-enterprise-files-us/left-hook/file/Q4tNvMWaoeSdCkDLAkQD.png?uploadId=SijJNrELtTU7osw9sjLjT3ATqo5TMDKka4TIhGXc6Bb_wp4yACoxUENVJuWfe9jlv6PNlWToiFZTF2HTwjdf2FMo71.f1e7ns5lWc4hGXLA73vhmYBqTwhuBEIS02rve&partNumber=1&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIARW2DEZZ4JVAXH74Y%2F20230628%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230628T162133Z&X-Amz-SignedHeaders=host&X-Amz-Expires=518400&X-Amz-Signature=49f4d714ce9b659a275120add169a473aecd41655e5efb772775104adda71e21',
      status: 200,
      statusText: 'OK',
      headers: [Headers],
      counter: 0
    }
  }
]
```

Reply:
@roboli I'm ok returning that. Then the user can check if it was a success, error, etc.